src/core/thread/mle.cpp

```diff
@@ -1662,6 +1669,18 @@ void Mle::HandleStateChanged(otChangedFlags aFlags)
         }
     }
 
+    if (aFlags & (OT_CHANGED_THREAD_BACKBONE_ROUTER_STATE | OT_CHANGED_THREAD_DOMAIN_PREFIX))
+    {
+        Get<ThreadNetif>().UnsubscribeMulticast(mAllDomainBackboneRouters);
```

Comment:
Can we unexpectedly remove other multicast addresses (e.x. `All Network Backbone Routers Multicast Address`) here, even though the chance is low. 
In the current implementation, `mAllDomainBackboneRouters` will contain the network prefix of last domain prefix when the device had BBR service enabled. When BBR service is turned off, `mAllDomainBackboneRouters` never change. Whenever this code runs, it will try to remove `mAllDomainBackboneRouters` which could equal to other subscribed multicast address. 
For example, if the current mesh local prefix equals to the old domain prefix set in `mAllDomainBackboneRouters` (a totally valid prefix setting), this code will unsubscribe `All Network Backbone Routers Multicast Address`.

I suggest a simple fix that we always set up `mAllDomainBackboneRouters` as long as `Get<BackboneRouter::Leader>().HasDomainPrefix()`.

Reply:
I don't think it will remove other irrelevant multicast addresses unexpectedly given that:
`mMulticastAddresses` is a linked-list, when [Netif::UnsubscribeMulticast()](https://github.com/openthread/openthread/blob/master/src/core/net/netif.cpp#L321), it finally goes to [LinkedList::Remove(const Type &aEntry)](https://github.com/openthread/openthread/blob/master/src/core/common/linked_list.hpp#L291) which accepts the reference as the formal parameter.

Reply:
Won't have this issue if it's comparing by reference. :+1: